### PR TITLE
RFC: remove clientversion.h usage from kernel

### PR DESCRIPTION
**Pull request by theuni**
RFC because this is rather hackish and I'd like to know if anyone has a better idea.

I'm working on trimming out headers that libbitcoinkernel should not export. To be clear: the problem is not what headers libbitcoinkernel itself uses, but what headers it requires users to use. Basically.. the headers used directly and indirectly by `bitcoin-chainstate.cpp`.

Topping the list of headers we definitely don't want to require for users are `bitcoin-config.h` and `clientversion.h`. The former is somewhat complicated, but `clientversion.h` only comes from one place: `dbwrapper.h`.

This PR gets rid of that include by creating a new constructor that sets the version in the cpp file, working around the problem.

Another potential solution would be to create a `MakeClientVersionCDataStream()` function which just creates a stream and sets the version, and relies on the caller to do the rest.

Any better ideas? Otherwise I'll undraft and ask for review.

### CODE REVIEW

**Comment by theuni** [comment]
Ping @TheCharlatan @MarcoFalke @ryanofsky

**Comment by TheCharlatan** [comment]
Concept ACK

**Review by bitcoinfinancier — APPROVED** [review]

**Comment by hebasto** [comment]
> Topping the list of headers we definitely don't want to require for users are `bitcoin-config.h` and `clientversion.h`.

Concept ACK on that.

**Comment by ajtowns** [comment]
Sorry if this is naive, but did you consider just introducing a `KERNEL_VERSION` constant that only gets bumped when it results in meaningful changes (like `PROTOCOL_VERSION` for p2p)? But @sipa's suggestion seems better if it's feasible.

**Comment by theuni** [comment]
> Sorry if this is naive, but did you consider just introducing a `KERNEL_VERSION` constant that only gets bumped when it results in meaningful changes (like `PROTOCOL_VERSION` for p2p)? But @sipa's suggestion seems better if it's feasible.

@ajtowns Thing is, `KERNEL_VERSION` would still have a purely internal meaning in this case (and one unrelated to the kernel version :), it would just not be tied to clientversion anymore. That approach would allow a client application to change the value in the header and potentially get some different serialization behavior in their app. Admittedly in this case, from what I can tell, the version is serialized but unused so that's not a real problem.

The approach in this PR which hard-codes the values into the binary so that the downstream user can't possibly affect behavior.

But yes, I'm totally up for dropping this and switching to serialization params if that helps with the dependency here.

**Comment by theuni** [comment]
Hmm, maybe there's a much simpler solution...

With the offending `CDataStream`s turned into `DataStream`s, compile fails in 2 places because version is used:

https://github.com/bitcoin/bitcoin/blob/33da5d0eb155993ab0d07c95704a0dd9e981c100/src/chain.h#L404-L419
https://github.com/bitcoin/bitcoin/blob/33da5d0eb155993ab0d07c95704a0dd9e981c100/src/primitives/block.h#L128-L134

Both of those, as far as I can tell, are reading/writing `nVersion` as a dummy value. The version used by streams is inconsistent.. sometimes it's CLIENT_VERSION, sometimes it's PROTOCOL_VERSION...

Is there any reason to keep the `GetVersion()` calls for these two at all since we're not using it at all? As far as I can tell it'd be safe to replace them with zeroes.

File: src/dbwrapper.h [hunk 1]
```diff
@@ -6,7 +6,6 @@
 #define BITCOIN_DBWRAPPER_H
 
 #include <attributes.h>
-#include <clientversion.h>
```
[the remark this answers is not in the document]
Comment (theuni) [reply src/dbwrapper.h:9]: FWIW I hacked this up here, just to see how the list is looking: https://github.com/theuni/bitcoin/commit/60da4fa6190e3a85d094f97dc8e578aeabcd22bf#diff-4cb884d03ebb901069e4ee5de5d02538c40dd9b39919c615d8eaa9d364bbbd77R1002-R1096

**Comment by fanquake** [comment]
Note that #25284 has now been merged.

**Comment by theuni** [comment]
> Note that #25284 has now been merged.

Great, thanks for the ping. Going to close this and rework after #25284 since the approach will be completely different now.

**Pull request closed without merging**